keps/sig-node/2837-pod-level-resource-spec/README.md

```diff
@@ -0,0 +1,1525 @@
+1. Directly from Pod-Level Resources (pod.spec.resources)
+   
+   If the pod.spec.resources is defined, the scheduler will use the resource
+   requests at the pod-level as the sole indicator of the pod's resource needs.
+   This includes the combined resource requirements of all containers within the
+   pod, including init containers, sidecar containers, and regular containers.
+
+2. Indirectly from Container-level Resources
+   
+   If the pod.spec.resources is NOT defined, the scheduler will derive the pod's
+   resource needs by aggregating the requests and limits of all containers (init,
+   sidecar and regular containers) within the pod.
+
+3. Both Pod-level and Container level
+   
+   If both pod-level and container-level resources are specified, the scheduler
+   will prioritize the pod-level resources and ignore the container-level
+   requests for node placement decisions.
+
+Note: It is important to consider how this change will affect observability tools
+(like kube-state-metrics) that monitor and report on Kubernetes' resource usage.
+To ensure these tools work correctly with pod-level resource specifications,
+these changes will be clearly documented to spread awareness around how these
+tools need to be updated before this feature is enabled by default in Beta.
+
+To further aid in this transition, the Sig Scheduling community is working on developing
+PodRequests library function that these tools can use to accurately retrive pod
+resource requests, taking into account both pod-level and container-level
+specifications. This effort is tracked in https://github.com/kubernetes/kubernetes/pull/124609
+
+#### QoS Changes
+
+With the addition of pod-level resource specifications, Kubernetes will
+prioritize those values when determining the Quality of Service (QoS) for a pod.
+Here's the order of preference:
+1. Pod Level Resources: If the resources are specified at the pod level,
+   Kubernetes will use those values to determine the QoS class.
+2. Container Level Resources: If pod-level resources are not specified,
+   Kubernetes will back to the resources defined at the container level for QoS
+   classficiation. The value of resources specified in the container’s
+  
+Essentially, pod-level resource specifications, when present, take precedence
+over container-level specifications in determining the QoS class of a pod.
+
+For each core resource type (CPU, Memory), a pod may have pod-scoped resource
+request and/or limit. Pods which have a defined request or limit for a given
+resource are said to "have explicitly resourced" or "be explicitly resourced".
+Pods which DO NOT have a defined request or limit for a given resource type are
+said to "have implicit resources" or "be implicitly resourced"
+
+* **Guaranteed QoS**
+
+    A pod is considered `Guaranteed` if, for each resource type (CPU and memory),
+    one of following conditions is true:
+    * It is explicitly resourced for that resource type AND the the pod's request
```

Comment:
one interesting side effect from this will be that the definition of "Guaranteed" will change a bit. Today guaranteeed indicates both - kuberentes side and app side. K8s side: there is always as much resources as you requested and there are no noisy neighbours who can cause your Pod to be evicted. App side: if app uses exact resources specified as a limit on a cgroup - it will always run fine. With the change - we are keeping the k8s side of it regarding noisy neighbours. But the app side of it is not true any longer. Containers will need to coordinate usage to not step into each other's limits.

Reply:
I believe we discussed this at SIG Node design review a while back. I disagree with the comment. 

Please note that pod-level resources are optional. This means the existing behavior of Guaranteed QoS with container-level guarantees remains fully intact. The new feature extends the capabilities, offering more flexibility without taking away existing functionalities.

Reply:
I think this is actually quite intuitive:
 - If you specify resources at the container level, it's guaranteed at the container level
 - If you specify resources at the pod level, it's guaranteed at the pod level
 
And like Dawn mentioned, it doesn't change the semantics today. If we want to evolve the QoS scope and definition in the future, this (container vs pod level) may come into play. We'll just need to be mindful of it.